toc/rfc/rfc-draft-pcap-bosh.md

```diff
@@ -0,0 +1,59 @@
+
+Since then, the project has evolved into a BOSH use case and a Cloud Foundry use case. This RFC discusses a proposal to get the BOSH use case integrated with BOSH director and BOSH cli.
+
+The main goal of the pcap feature is to allow users (BOSH operator in this case) to easily capture network traffic across multiple VMs by specifying various parameters in PCAP CLI:
+- BOSH deployment
+- BOSH instance group(s)
+- BPF [Berkeley Packet Filter](https://www.tcpdump.org/manpages/pcap-filter.7.html) filter to apply
+
+PCAP CLI will form a capture request and send it to PCAP API. The capture will then take place in real-time similar to running [tcpdump](https://www.tcpdump.org/manpages/tcpdump.1.html) in a distributed fashion across multiple machines. The PCAP API will contact PCAP AGENT on all nodes involved and trigger a network capture. The PCAP AGENTs will concurrently stream all captured traffic back to PCAP AGENT where all streams are merged into one contiguous stream that is sent back to the PCAP CLI where it is then stored as a PCAP FILE on disk.
+
+Trust is established as follows:
+1. PCAP CLI uses the BOSH CLI `access_token` and sends them to PCAP API
+2. PCAP API verifies `access_token` against BOSH UAA and establishes mTLS with PCAP AGENT
+3. PCAP AGENT only accepts mTLS connections from PCAP AGENT via a trusted CA
+
+## Problem
+
+The current architecture for "PCAP for BOSH" is depicted as follows:
+
+![PCAP BOSH Architecture](rfc-draft-pcap-bosh/tcpdump-for-bosh.svg)
+
+
+There are several flaws with this architecture that are the result of missing integration:
+1. The PCAP BOSH CLI does not share context with BOSH CLI (because unlike CF CLI, BOSH CLI has no plugin mechanism).
+2. PCAP BOSH CLI reads the user's `access_token` and `refresh_token`, talk to BOSH UAA to get a new `access_token` using the `refresh_token` and write both back to `.bosh/config` - a file which is also accessed by BOSH CLI.
+3. The `access_token` is then sent to PCAP API which uses it as a credential to talk to BOSH DIRECTOR
+4. PCAP API needs to speak to BOSH DIRECTOR in order to find the endpoints (IP:PORT) of each PCAP AGENT needed for the capture.
+5. In order to make this feature work, PCAP AGENT needs to be available on all BOSH managed VMs. The only way today would be to use it as an [addon](https://bosh.io/docs/runtime-config/#addons)
+
+## Proposal
+
+### Integrate PCAP CLI as a "pcap" keyword into BOSH CLI
+
+- This would allow PCAP BOSH CLI to share context with BOSH CLI and have access to the user's login session as well as the currently used BOSH environment.
+
+- Another benefit would be that users would not have to download the PCAP BOSH CLI separately. 
+
+### Co-locate PCAP AGENT with BOSH AGENT
+
+- This would make network capture a "built-in" feature of BOSH. It removes the need to manually add PCAP AGENT as a job to deployment manifests or configure it as an addon.
```

Comment:
On agent side the proposal is to have the `pcap` functionality integrated directly into the BOSH agent by importing our library. Unless there is a compelling reason not to do so?

Reply:
Fair point, I didn't know they rewrote it from ruby to go already. I will rephrase to "integrate with BOSH AGENT".
The details of the integration (submodule or not etc.) will be discussed with the community. We only need to maintain the ability to also have pcap-agent source on a separate repo so that it may be integrated with CF also.

Who knows, could also become a separate "pcap-bosh-agent" if integrating while retaining original sources is not a feasible option.

Reply:
We will not split the source. All go code related to pcap will remain in a single repository (and go module), in this case we would probably just have a dependency to that module from the bosh-agent.